File: craftcv/backend/app/schemas/template.py
from datetime import datetime
from pydantic import BaseModel, Field
from typing import Any, Optional

from app.core.validation import TemplateSchemaValidator


class TemplateField(BaseModel):
    type: str
    label: str = Field(max_length=255)
    required: bool = False
    max_length: int | None = None
    min: int | None = None
    max: int | None = None
    options: list[str] | None = None


class SectionDefinition(BaseModel):
    id: str = Field(pattern=r"^[a-z0-9-]+$", max_length=64)
    type: str
    label: str = Field(max_length=255)
    required: bool = False
    min_items: int = 0
    max_items: int | None = None
    fields: dict[str, TemplateField]


class LayoutDefinition(BaseModel):
    columns: int = 1
    color_scheme: dict[str, str] | None = None
    fonts: dict[str, str] | None = None


class TemplateDefinition(BaseModel):
    sections: list[SectionDefinition]
    layout: LayoutDefinition


class TemplateCreate(BaseModel):
    name: str
    description: str | None = None
    definition: Any
    default_styles: Any | None = None
    is_public: bool = False


class TemplateResponse(BaseModel):
    id: str
    name: str
    description: str | None = None
    thumbnail_url: str | None = None
    is_public: bool
    definition: Any
    default_styles: Any
    created_at: datetime
    updated_at: datetime

File: craftcv/frontend/src/types/template.ts
export interface TemplateField {
  type: "text" | "url" | "date" | "boolean" | "richtext" | "list" | "tags" | "select";
  label: string;
  required?: boolean;
  max_length?: number;
  min?: number;
  max?: number;
  options?: string[];
}

export interface SectionDefinition {
  id: string;
  type: string;
  label: string;
  required?: boolean;
  min_items?: number;
  max_items?: number;
  fields: Record<string, TemplateField>;
}

export interface LayoutDefinition {
  columns: number;
  color_scheme: Record<string, string>;
  fonts: Record<string, string>;
}

export interface TemplateDefinition {
  sections: SectionDefinition[];
  layout: LayoutDefinition;
}

export interface Template {
  id: string;
  name: string;
  description?: string;
  thumbnail_url?: string;
  is_public: boolean;
  definition: TemplateDefinition;
  default_styles: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}
